### PR TITLE
python39Packages.functorch: init at 0.1.1

### PR DESCRIPTION
**Pull request by samuela**
###### Description of changes

See https://twitter.com/PyTorch/status/1528800506067288065.

cc @SomeoneSerge might also be interested?

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/functorch/default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  # Somewhat surprisingly pytorch is actually necessary for the build process.
+  # `setup.py` imports `torch.utils.cpp_extension`.
+  nativeBuildInputs = [
+    ninja
+    pytorch
+    which
+  ];
+
+  # Tests require expecttest which is not yet packaged as of 2022-05-24.
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/functorch/default.nix:29]: Might be worth packaging. Unfortunately, it comes up rather regularly, I even keep an [expression](https://github.com/SomeoneSerge/pkgs/blob/master/pkgs/ezy-expecttest.nix) for it

**Comment by SomeoneSerge** [comment]
Result of `nixpkgs-review pr 174248` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python310Packages.functorch</li>
    <li>python39Packages.functorch</li>
  </ul>
</details>

File: pkgs/development/python-modules/functorch/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/functorch/default.nix:29]: That looks rather easy to package.

Comment (samuela) [reply pkgs/development/python-modules/functorch/default.nix:36]: Done in https://github.com/NixOS/nixpkgs/pull/174248/commits/b63b1f2266b7d748002e97566bf45b8f5f17d51f

**Comment by samuela** [comment]
Ok, I packaged expecttest now and tried enabling the test suite but functorch is failing the test suite with errors that look like
```
______________________ ERROR collecting test/test_ops.py _______________________
ImportError while importing test module '/build/source/test/test_ops.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/nix/store/vzqny68wq33dcg4hkdala51n5vqhpnwc-python3-3.9.12/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test/test_ops.py:19: in <module>
    from common_utils import (
test/common_utils.py:9: in <module>
    import functorch
functorch/__init__.py:7: in <module>
    from . import _C
E   ImportError: cannot import name '_C' from partially initialized module 'functorch' (most likely due to a circular import) (/build/source/functorch/__init__.py)
```
I tried adding
```nix
  preCheck = ''
    export PYTHONPATH=$out/${python.sitePackages}:$PYTHONPATH
  '';
```
but to no avail.

Any ideas?

**Comment by SomeoneSerge** [comment]
It's not particularly clean, but:

```nix
  preCheck = ''
    rm -rf functorch
  '';
```

**Comment by samuela** [comment]
Ok, switched expecttest to use `fetchFromGitHub` thanks to https://github.com/ezyang/expecttest/issues/3 being resolved and disabled the functorch tests according to the guidance from https://github.com/pytorch/functorch/issues/835.

Everything builds and should be ready for review now!

File: pkgs/development/python-modules/expecttest/default.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+  src = fetchFromGitHub {
+    owner = "ezyang";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-5CnpVFSbf3FcAa06Y7atG8sxu8uevpfrliB2HuVcrx0=";
+  };
+
+  buildInputs = [ poetry ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/expecttest/default.nix:20]: Missing pythonImportsCheck.
Can we run https://github.com/ezyang/expecttest/blob/main/test_expecttest.py ?

Comment (samuela) [reply pkgs/development/python-modules/expecttest/default.nix:22]: Oh good catch, fixed in https://github.com/NixOS/nixpkgs/pull/174248/commits/4edbc05a5eb5124822bd2066f85ec5abfa452f54

**Comment by SuperSandro2000** [comment]
@ofborg eval

**Comment by SuperSandro2000** [comment]
aarch64-linux fails to build

``` 
during RTL pass: reload
In file included from /build/source/aten/src/ATen/Parallel.h:153,
                 from /build/source/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:11,
                 from /build/source/build/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp.DEFAULT.cpp:1:
/build/source/aten/src/ATen/ParallelOpenMP.h: In function ‘_ZN2at8internal15invoke_parallelIZNS_3vml7DEFAULT7vlgammaIfEEvPT_PKS5_lEUlllE_EEvlllRS7_._omp_fn.0’:
/build/source/aten/src/ATen/ParallelOpenMP.h:22:9: internal compiler error: Max. number of generated reload insns per insn is achieved (90)
   22 | #pragma omp parallel
      |         ^~~
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://gcc.gnu.org/bugs/> for instructions.
[3063/3450] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/RegisterCPU.cpp.ocpp.o.cpp.o.o.oon.cpp.o
ninja: build stopped: subcommand failed.
```

**Comment by samuela** [comment]
Interesting, my guess is that functorch doesn't support aarch64-linux. I'll update meta.platforms accordingly

**Comment by samuela** [comment]
Done in https://github.com/NixOS/nixpkgs/pull/174248/commits/65450bdd42b9132d5633245a44e16f12e51820a6

**Comment by samuela** [comment]
If no one objects I'll go ahead and merge tomorrow

**Pull request merged**